e2e2/test/images/bert-training/train.py

```diff
@@ -0,0 +1,100 @@
+    input_ids, labels = mask_tokens(tokenized_inputs.input_ids, tokenizer, mlm_probability)
+
+    # NSP task: create dummy pairs
+    next_sentence_labels = torch.randint(0, 2, (num_samples,))
+
+    return TensorDataset(input_ids, tokenized_inputs.attention_mask, labels, next_sentence_labels)
+
+def mask_tokens(inputs, tokenizer, mlm_probability):
+    labels = inputs.clone()
+    probability_matrix = torch.full(labels.shape, mlm_probability)
+    special_tokens_mask = [
+        tokenizer.get_special_tokens_mask(val, already_has_special_tokens=True) for val in labels.tolist()
+    ]
+    probability_matrix.masked_fill_(torch.tensor(special_tokens_mask, dtype=torch.bool), value=0.0)
+    masked_indices = torch.bernoulli(probability_matrix).bool()
+    labels[~masked_indices] = -100  # We only compute loss on masked tokens
+
+    inputs[masked_indices] = tokenizer.convert_tokens_to_ids(tokenizer.mask_token)
+
+    return inputs, labels
+
+def setup(rank, world_size):
+    os.environ['MASTER_ADDR'] = os.environ['MASTER_ADDR']  # Kubernetes sets this
+    os.environ['MASTER_PORT'] = os.environ['MASTER_PORT']  # Kubernetes sets this
+    dist.init_process_group("nccl", rank=rank, world_size=world_size)
+    torch.cuda.set_device(rank)
+    print(f"Process {rank} initialized, using GPU {rank}")
+
+def cleanup():
+    dist.destroy_process_group()
+
+def train_bert(rank, world_size, model, tokenizer):
+    setup(rank, world_size)
+   
+    tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
+    model = BertForPreTraining.from_pretrained('bert-base-uncased').to(rank)
+    ddp_model = DDP(model, device_ids=[rank])
+
+    dataset = create_dummy_data(tokenizer)
+    train_sampler = torch.utils.data.distributed.DistributedSampler(dataset, num_replicas=world_size, rank=rank)
+    train_dataloader = DataLoader(dataset, sampler=train_sampler, batch_size=8)
+
+    optimizer = torch.optim.AdamW(ddp_model.parameters(), lr=0.001)
+    criterion = torch.nn.CrossEntropyLoss()
+
+    start_time = time.time()
+
+    for epoch in range(1):  # Short run for testing
+        ddp_model.train()
+        for batch in train_dataloader:
+            optimizer.zero_grad()
+            inputs, masks, labels, next_sentence_labels = batch
+            inputs, masks, labels, next_sentence_labels = inputs.to(rank), masks.to(rank), labels.to(rank), next_sentence_labels.to(rank)
+            outputs = ddp_model(input_ids=inputs, attention_mask=masks, labels=labels, next_sentence_label=next_sentence_labels)
+            loss = outputs.loss
+            loss.backward()
+            optimizer.step()
+
+    end_time = time.time()
+    training_time = end_time - start_time
+    throughput = len(dataset) / training_time
+
+    metrics = {
+        "training_time": training_time,
+        "throughput": throughput,
+    }
+
+    print(f"Process {rank} - Training time: {training_time:.2f} seconds")
+    print(f"Process {rank} - Throughput: {throughput:.2f} samples/second")
```

Comment:
do we need to dump this output to disk so we can use to upload to s3?

Reply:
Potentially... I also was considering of writing directly to s3 as well, but was curious to hear other's perspective(s). My intuition says writing to S3 is the long-term solution (once a stable schema is solidified), but short term just doing something like writing to disk or stdout might be the way to go.

Reply:
i agree it can go to s3, cloudwatch, or etc once we know where this is going. we should have this output for sure.

Reply:
it should be fine dump to disk for short term, and you have enough flexibility to play POC with different long-term destinations

Reply:
@Issacwww  Are there any concerns/considerations with writing from the container to the host machine?

Reply:
oh, good call out, this is on the tod worker, dump to disk has no difference between stdout... so stdout should be fine now.